Job op

Noremac201 opened the pull request:
Make sure you have checked _all_ steps below.

### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-XXX
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.


### Description
- [ ] Here are some details about my PR, including screenshots of any UI changes:


### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:


### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


### Documentation
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.


### Code Quality
- [ ] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

TrevorEdwards commented on airflow/contrib/operators/gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import subprocess
```
Is there a reason this is separate from lines 21/22?

TrevorEdwards commented on airflow/contrib/operators/gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+        Executes a task in a Kubernetes pod in the specified Google Kubernetes Engine
+        cluster
+
+        This Operator assumes that the system has gcloud installed and has default
```
Might be helpful to link to https://cloud.google.com/docs/authentication/production#providing_credentials_to_your_application

TrevorEdwards commented on airflow/contrib/operators/gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+        cluster, etc.
+
+        This does not require a gcp_conn_id nor api_version since it does not end up
+        getting a connection from a hook, nor using a specific api version
```
Not sure if these details are necessary, but feel free to leave in if you want.

I would also explicitly document that this executes as a KubernetesPodOperator once it obtains credentials and accepts all of its arguments.

TrevorEdwards commented on airflow/contrib/operators/gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+
+            # Tell KubernetesPodOperator where the config file is
+            self.config_file = os.environ[KUBE_CONFIG_ENV_VAR]
+            super(GKEPodOperator, self).execute(context)
```
Just want to confirm - does tempfile clean itself up at the end?

Noremac201 replied:
The file is removed after the file is closed.

TrevorEdwards commented on tests/contrib/operators/test_gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,82 @@
+# specific language governing permissions and limitations
+# under the License.
+import tempfile
+import unittest
```
Similar to other file, is there a reason you choose this import formatting/order?

TrevorEdwards commented on tests/contrib/operators/test_gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,82 @@
+from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator
+
+try:
+    from unittest import mock
```
is this importing mock style standard for Airflow?

Noremac201 replied:
As far as I can tell, yes.

TrevorEdwards commented on airflow/contrib/operators/gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,100 @@
+
+            operator = GKEPodOperator(task_id='pod_op',
+                                      project_id='my-project',
+                                      location='my-location',
```
Maybe more realistic location for location- would 'us-central1-a' work?

TrevorEdwards commented on airflow/contrib/operators/gcp_kubernetes_operator.py:
```diff
@@ -0,0 +1,100 @@
+                                      cluster_name='my-cluster-name',
+                                      name='task-name',
+                                      namespace='default',
+                                      image='bash')
```
is 'bash' an actual image?

TrevorEdwards reviewed (approved):
A couple comments, then LGTM

Squash and PR?

Noremac201 committed:
[AIRFLOW-2658] Add GCP specific k8s pod operator

Executes a task in a Kubernetes pod in the specified Google Kubernetes
Engine cluster. This makes it easier to interact with GCP kubernetes
engine service because it encapsulates acquiring credentials.
